### ISSUE TITLE
Module "fs" has been externalized for browser compatibility and cannot be accessed in client code

### ISSUE THREAD

**Issue by tjx666** [xiaoxiangmoe/vite-plugin-commonjs-externals#4]
vite config:

```typescript
import vue from '@vitejs/plugin-vue';
import { defineConfig } from 'vite';
import ViteComponents from 'vite-plugin-components';
import commonjsExternals from 'vite-plugin-commonjs-externals';

import gdAntdPlusResolver from './utils/gd-ant-plus-resolver';
import builtinModules from './utils/builtin-modules';

const commonjsPackages = [
    'electron',
    'electron/main',
    'electron/common',
    'electron/renderer',
    ...builtinModules,
] as const;

// https://vitejs.dev/config/
export default defineConfig({
    plugins: [
        vue(),
        ViteComponents({
            dirs: ['components'],
            globalComponentsDeclaration: './typings/components.d.ts',
            customComponentResolvers: [
                gdAntdPlusResolver({
                    importLess: true,
                    resolveIcons: true,
                }),
            ],
        }),
        commonjsExternals({ externals: commonjsPackages }),
    ],
    css: {
        preprocessorOptions: {
            less: {
                javascriptEnabled: true,
            },
        },
    },
});
```

vue code:

```vue
<template>
    <div class="settings">
        <div class="settings-item">
            <label for="aeInstallationDirectory">AE 安装目录：</label>
            <span>{{ aeSdkInstallDir }}</span>
            <a-button>选择</a-button>
        </div>
    </div>
</template>

<script lang="ts" setup>
import path from 'path';
import fs from 'fs';
import util from 'util';
import { onMounted, ref } from 'vue';

console.log(fs.readFileSync);
const access = util.promisify(fs.access);
const readDir = util.promisify(fs.readdir);

async function existsFile(path: string) {
    try {
        await access(path, fs.constants.F_OK);
    } catch (error) {
        return false;
    }
    return true;
}

async function getAESdkInstallationDir() {
    const defaultApplicationDir = '/Applications';
    if (await existsFile(defaultApplicationDir)) {
        const files = await readDir(defaultApplicationDir);
        for (const file of files) {
            if (file.match(/Adobe After Effects 20\d\d/)) {
                return path.resolve(defaultApplicationDir, file);
            }
        }
    }
    return '';
}

const aeSdkInstallDir = ref('');

onMounted(async () => {
    // aeSdkInstallDir.value = await getAESdkInstallationDir();
});
</script>

<style lang="less">
.settings {
    height: 100vh;

    // &-item {

    // }
}
</style>
```

version info:

"electron": "^13.1.6"

![image](https://user-images.githubusercontent.com/41773861/126141371-231a1569-3d4e-40fd-9f2e-c6c929542c1f.png)

![image](https://user-images.githubusercontent.com/41773861/126141194-65b6eb86-437b-45c4-be7f-0954566b3f3a.png)

**Comment by tjx666** [comment]
好像 vite 就直接禁止访问浏览器环境外的代码 https://github.com/vitejs/vite/blob/f623ba37dbc517656a7ffb10bc1b3cb221a4bc72/packages/vite/src/node/plugins/resolve.ts#L255

**Comment by cll123456** [comment]
怎么解决这个问题呢？@tjx666

**Comment by cll123456** [comment]
如果我在内部就要使用fs,path等模块呢

**Comment by tjx666** [comment]
我们的项目已经不在浏览器代码直接访问 node 代码了，通过 preload script 暴露 ipcRender 给浏览器代码，前端代码需要访问 node 的时候通过 ipcRender 和 主进程通信。

参考： https://github.com/tjx666/electron-vue-vite-boilerplate

**Comment by beorn** [comment]
You also have to set ```config.optimizeDeps.exclude``` ([ref](https://vitejs.dev/config/#optimizedeps-exclude)) to the modules you want vite're pre-bundle dependency system to ignore (and therefore pass through to electron/node to deal with as a require).  Maybe this plugin should do that automatically, or have an option to do so?

**Comment by q1971595727** [comment]
crypto 也抛这个错。无语死

**Comment by tjx666** [comment]
try another plugin: [vite-plugin-optimizer](https://github.com/caoxiemeihao/vite-plugins/tree/main/packages/optimizer)

```javascript
import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';
import optimizer from 'vite-plugin-optimizer';

export default defineConfig({
    plugins: [
        vue(),
        optimizer({
            path: () => ({
                find: /^path$/,
                code: `const path = require('path'); export { path as default }`,
            }),
        }),
    ],
});

```

**Comment by tjx666** [comment]
还有一种思路，直接在 index.html 里面吧模块全挂到 window.node 对象上，用的地方直接通过 window.node 对象访问。

**Comment by yeminxuan** [comment]
这个做法会导致vite报错无法使用require，这个我尝试了`vite-plugin-require-transform`也没用

**Comment by awvalenti** [comment]
This idea may help:
https://stackoverflow.com/questions/73335430/module-util-has-been-externalized-for-browser-compatibility-cannot-access-ut
Something like this:
```js
export default defineConfig({
  resolve: {
    alias: {
      util: 'util/',
    },
  },
})
```

**Comment by dusionlike** [comment]
刚遇到，过来说下，应该是默认配置里没有包含vue文件。
https://github.com/xiaoxiangmoe/vite-plugin-commonjs-externals/blob/6d895f4b91f1668b15df2ef4701a9bdd412c24b7/src/index.ts#L128-L131

在 `vite.config.ts` 加上即可

```diff
commonjsExternals({
    externals,
+    exts: ['ts', 'tsx', 'js', 'jsx', 'mjs', 'cjs', 'vue'],
}),
```

**Comment by xiaoxiangmoe** [comment]
released in [0.1.3](https://www.npmjs.com/package/vite-plugin-commonjs-externals/v/0.1.3)
